Issues compiling on ARM Macs

When trying to compile for release on an ARM mac, issues get thrown about architectures, as the xcode compiler tries to link ARM libraries to x86 binaries.

```
Ld /Users/anze/d/OpenSCToken/build/OpenSCTokenApp.build/Release/opensc-pkcs11.build/Objects-normal/x86_64/Binary/opensc-pkcs11.dylib normal x86_64 (in target 'opensc-pkcs11' from project 'OpenSCTokenApp')
    cd /Users/anze/d/OpenSCToken
    /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -target x86_64-apple-macos10.13 -dynamiclib -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.1.sdk -L/Users/anze/d/OpenSCToken/build/EagerLinkingTBDs -L/Users/anze/d/OpenSCToken/build/Release -F/Users/anze/d/OpenSCToken/build/EagerLinkingTBDs -F/Users/anze/d/OpenSCToken/build/Release -filelist /Users/anze/d/OpenSCToken/build/OpenSCTokenApp.build/Release/opensc-pkcs11.build/Objects-normal/x86_64/opensc-pkcs11.dylib.LinkFileList -Xlinker -object_path_lto -Xlinker /Users/anze/d/OpenSCToken/build/OpenSCTokenApp.build/Release/opensc-pkcs11.build/Objects-normal/x86_64/opensc-pkcs11.dylib_lto.o /Users/anze/d/OpenSCToken/OpenSC/src/libopensc/.libs/libopensc.a /Users/anze/d/OpenSCToken/build/openpace/lib/libeac.a /Users/anze/d/OpenSCToken/build/openssl/lib/libcrypto.a -framework CryptoTokenKit -framework Cocoa -lz -Xlinker -no_adhoc_codesign -Xlinker -dependency_info -Xlinker /Users/anze/d/OpenSCToken/build/OpenSCTokenApp.build/Release/opensc-pkcs11.build/Objects-normal/x86_64/opensc-pkcs11.dylib_dependency_info.dat -o /Users/anze/d/OpenSCToken/build/OpenSCTokenApp.build/Release/opensc-pkcs11.build/Objects-normal/x86_64/Binary/opensc-pkcs11.dylib
ld: warning: ignoring file /Users/anze/d/OpenSCToken/OpenSC/src/libopensc/.libs/libopensc.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /Users/anze/d/OpenSCToken/build/openpace/lib/libeac.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm64
ld: warning: ignoring file /Users/anze/d/OpenSCToken/build/openssl/lib/libcrypto.a, building for macOS-x86_64 but attempting to link with file built for macOS-arm6
```

The issue itself seems to be fixed, but now I get the following:
```
/Users/anze/d/openarm/OpenSC/src/pkcs11/framework-pkcs15.c:3268:32: error: implicit conversion loses integer precision: 'CK_RV' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
                        return sc_to_cryptoki_error(rv, "C_GenerateKeyPair");
                               ~~~~~~~~~~~~~~~~~~~~ ^~
/Users/anze/d/openarm/OpenSC/src/pkcs11/framework-pkcs15.c:6302:7: error: implicit conversion loses integer precision: 'CK_RV' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
        rc = sc_pkcs11_register_mechanism(p11card, mt, NULL);
           ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/framework-pkcs15.c:6310:7: error: implicit conversion loses integer precision: 'CK_RV' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
        rc = sc_pkcs11_register_mechanism(p11card, mt, NULL);
           ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/framework-pkcs15.c:6318:7: error: implicit conversion loses integer precision: 'CK_RV' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
        rc = sc_pkcs11_register_mechanism(p11card, mt, NULL);
           ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```
/Users/anze/d/openarm/OpenSC/src/pkcs11/mechanism.c:109:12: error: implicit conversion loses integer precision: 'CK_RV' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
                && (rv = mt->copy_mech_data(mt->mech_data, (void **) &(*new_mt)->mech_data)) != CKR_OK) {
                       ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/mechanism.c:135:79: error: implicit conversion loses integer precision: 'CK_FLAGS' (aka 'unsigned long') to 'unsigned int' [-Werror,-Wshorten-64-to-32]
        if ((existing_mt = sc_pkcs11_find_mechanism(p11card, mt->mech, mt->mech_info.flags))) {
                           ~~~~~~~~~~~~~~~~~~~~~~~~                    ~~~~~~~~~~~~~~^~~~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/mechanism.c:165:12: error: implicit conversion loses integer precision: 'CK_RV' (aka 'unsigned long') to 'int' [-Werror,-Wshorten-64-to-32]
        if ((rv = sc_pkcs11_copy_mechanism(mt, &copy_mt)) != CKR_OK) {
                ~ ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/mechanism.c:573:49: error: implicit conversion loses integer precision: 'CK_ULONG' (aka 'unsigned long') to 'unsigned int' [-Werror,-Wshorten-64-to-32]
        rv = signature_data_buffer_append(data, pPart, ulPartLen);
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~              ^~~~~~~~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/mechanism.c:596:49: error: implicit conversion loses integer precision: 'CK_ULONG' (aka 'unsigned long') to 'unsigned int' [-Werror,-Wshorten-64-to-32]
                rv = signature_data_buffer_append(data, hash, len);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~             ^~~
/Users/anze/d/openarm/OpenSC/src/pkcs11/mechanism.c:832:55: error: implicit conversion loses integer precision: 'CK_ULONG' (aka 'unsigned long') to 'unsigned int' [-Werror,-Wshorten-64-to-32]
        CK_RV rv = signature_data_buffer_append(data, pPart, ulPartLen);
                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~              ^~~~~~~~~
6 errors generated.
```

--- a/OpenSCTokenApp.xcodeproj/project.pbxproj
+++ b/OpenSCTokenApp.xcodeproj/project.pbxproj
@@ -608,7 +608,6 @@
 					$SRCROOT/build/openssl/include/,
 				);
 				MACOSX_DEPLOYMENT_TARGET = 10.12;
-				ONLY_ACTIVE_ARCH = YES;
 				OTHER_LDFLAGS = (
 					$SRCROOT/OpenSC/src/libopensc/.libs/libopensc.a,
 					$SRCROOT/build/openpace/lib/libeac.a,

--- a/OpenSCTokenApp.xcodeproj/project.pbxproj
+++ b/OpenSCTokenApp.xcodeproj/project.pbxproj
@@ -3,7 +3,7 @@
 	archiveVersion = 1;
 	classes = {
 	};
-	objectVersion = 46;
+	objectVersion = 53;
 	objects = {
 
 /* Begin PBXBuildFile section */
@@ -407,6 +407,7 @@
 		2F3883901CF46C6E00F114A7 /* Project object */ = {
 			isa = PBXProject;
 			attributes = {
+				BuildIndependentTargetsInParallel = YES;
 				LastUpgradeCheck = 1100;
 				ORGANIZATIONNAME = "Libor sykora";
 				TargetAttributes = {
@@ -651,7 +652,10 @@
 			buildSettings = {
 				INFOPLIST_FILE = OpenSCTokenApp/Info.plist;
 				INSTALL_PATH = "$(LOCAL_APPS_DIR)/Utilities";
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/../Frameworks";
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/../Frameworks",
+				);
 				PRODUCT_BUNDLE_IDENTIFIER = "org.opensc-project.mac.opensctoken.OpenSCTokenApp";
 				PRODUCT_NAME = "$(TARGET_NAME)";
 			};
@@ -662,7 +666,10 @@
 			buildSettings = {
 				INFOPLIST_FILE = OpenSCTokenApp/Info.plist;
 				INSTALL_PATH = "$(LOCAL_APPS_DIR)/Utilities";
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/../Frameworks";
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/../Frameworks",
+				);
 				PRODUCT_BUNDLE_IDENTIFIER = "org.opensc-project.mac.opensctoken.OpenSCTokenApp";
 				PRODUCT_NAME = "$(TARGET_NAME)";
 			};
@@ -673,7 +680,11 @@
 			buildSettings = {
 				CODE_SIGN_ENTITLEMENTS = OpenSCToken/opensctoken.entitlements;
 				INFOPLIST_FILE = OpenSCToken/Info.plist;
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/../Frameworks @executable_path/../../../../Frameworks";
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/../Frameworks",
+					"@executable_path/../../../../Frameworks",
+				);
 				PRODUCT_BUNDLE_IDENTIFIER = "org.opensc-project.mac.opensctoken.OpenSCTokenApp.OpenSCToken";
 				PRODUCT_NAME = "$(TARGET_NAME)";
 			};
@@ -684,7 +695,11 @@
 			buildSettings = {
 				CODE_SIGN_ENTITLEMENTS = OpenSCToken/opensctoken.entitlements;
 				INFOPLIST_FILE = OpenSCToken/Info.plist;
-				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/../Frameworks @executable_path/../../../../Frameworks";
+				LD_RUNPATH_SEARCH_PATHS = (
+					"$(inherited)",
+					"@executable_path/../Frameworks",
+					"@executable_path/../../../../Frameworks",
+				);
 				PRODUCT_BUNDLE_IDENTIFIER = "org.opensc-project.mac.opensctoken.OpenSCTokenApp.OpenSCToken";
 				PRODUCT_NAME = "$(TARGET_NAME)";
 			};

--- a/bootstrap
+++ b/bootstrap
@@ -5,6 +5,12 @@ PREFIX=/
 export MACOSX_DEPLOYMENT_TARGET="10.12"
 RESOURCES=/Applications/Utilities/OpenSCTokenApp.app/Contents/PlugIns/OpenSCToken.appex/Contents/Resources
 
+xcode_ver=$(xcodebuild -version | sed -En 's/Xcode[[:space:]](.*)/\1/p')
+base_ver="12.2"
+if [ $(echo -e $base_ver"\n"$xcode_ver | sort -V | head -1) == "$base_ver" ]; then
+	export BUILD_ARM="true"
+fi
+
 set -ex
 
 if [ -z "$OPENSSL_CFLAGS" -a -z "$OPENSSL_LIBS" ] ; then
@@ -14,15 +20,30 @@ if [ -z "$OPENSSL_CFLAGS" -a -z "$OPENSSL_LIBS" ] ; then
 			git clone --depth=1 https://github.com/openssl/openssl.git -b OpenSSL_1_1_1-stable
 		fi
 		cd openssl
-		KERNEL_BITS=64 ./config --prefix=$PREFIX
-		make update
-		make depend
-		make
+		MACHINE=x86_64 ./config no-shared --prefix=$PREFIX
+		make clean
+		make -j 4
 		make DESTDIR=$BUILDPATH/openssl install_sw
+		if test -n "${BUILD_ARM}"; then
+			make clean
+			MACHINE=arm64 KERNEL_BITS=64 ./config no-shared --prefix=$PREFIX
+			make -j 4
+			make DESTDIR=$BUILDPATH/openssl_arm64 install_sw
+			lipo -create $BUILDPATH/openssl_arm64/$PREFIX/lib/libcrypto.a $BUILDPATH/openssl/$PREFIX/lib/libcrypto.a -output libcrypto.a
+			lipo -create $BUILDPATH/openssl_arm64/$PREFIX/lib/libssl.a $BUILDPATH/openssl/$PREFIX/lib/libssl.a -output libssl.a
+			mv libcrypto.a $BUILDPATH/openssl/$PREFIX/lib/libcrypto.a
+			mv libssl.a $BUILDPATH/openssl/$PREFIX/lib/libssl.a
+		fi
 		cd ..
 	fi
 	export OPENSSL_CFLAGS="`env PKG_CONFIG_PATH=$BUILDPATH/openssl/$PREFIX/lib/pkgconfig PKG_CONFIG_SYSROOT_DIR=$BUILDPATH/openssl pkg-config --static --cflags libcrypto`"
 	export OPENSSL_LIBS="`  env PKG_CONFIG_PATH=$BUILDPATH/openssl/$PREFIX/lib/pkgconfig PKG_CONFIG_SYSROOT_DIR=$BUILDPATH/openssl pkg-config --static --libs   libcrypto`"
+fi
+
+if test -n "${BUILD_ARM}"; then
+	export CFLAGS="$CFLAGS -arch x86_64 -arch arm64"
+	export LDFLAGS="$LDFLAGS -arch x86_64 -arch arm64"
+	export OBJCFLAGS=$CFLAGS
 fi
 
 if [ -z "$OPENPACE_CFLAGS" -a -z "$OPENPACE_LIBS" ] ; then

--- a/bootstrap
+++ b/bootstrap
@@ -49,7 +49,7 @@ fi
 if [ -z "$OPENPACE_CFLAGS" -a -z "$OPENPACE_LIBS" ] ; then
 	if ! test -e $BUILDPATH/openpace/$PREFIX/lib/pkgconfig; then
 		if ! test -e openpace; then
-			git clone --depth=1 https://github.com/frankmorgner/openpace.git -b 1.1.2
+			git clone --depth=1 https://github.com/frankmorgner/openpace.git -b 1.1.3
 		fi
 		cd openpace
 		if ! test -e configure; then

